feat(ras-acc): use newsletters subscription lists setting to manage post-checkout newsletter signup

chickenn00dle opened the pull request:
### All Submissions:

* [ ] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/trunk/.github/CONTRIBUTING.md)?
* [ ] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [ ] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request, and the reason for such changes. -->

Closes https://app.asana.com/0/1206943664367857/1205992141743582/f

These changes are paired with https://github.com/Automattic/newspack-blocks/pull/1727 and https://github.com/Automattic/newspack-newsletters/pull/1502

This PR:

Updates the Newsletter Subscription List setting copy to reflect the setting will also apply to checkout:

![Screenshot 2024-05-09 at 14 36 08](https://github.com/Automattic/newspack-plugin/assets/17905991/d024afed-eaf3-41f8-a3a6-1de61303b9ef)

Filters the post-checkout newsletter lists to include only lists that the reader is not signed up for and are not premium.

![Screenshot 2024-05-09 at 14 37 43](https://github.com/Automattic/newspack-plugin/assets/17905991/eb570960-cd2a-4672-a1ac-19e909fccb8a)

NOTE: Post-checkout newsletter signup form styles are being worked on in https://github.com/Automattic/newspack-plugin/pull/3091

### How to test the changes in this Pull Request:

Be sure to check out https://github.com/Automattic/newspack-blocks/pull/1727 and https://github.com/Automattic/newspack-newsletters/pull/1502 before testing.

1. Be sure to have at least two subscriptions lists, one that is associated with a WC membership plan, and one that is not.
2. Go to the Engagement reader activation wizard and disable the newsletter subscription lists setting: Newspack > Reader Revenue > Show Advanced Settings > Newsletter Subscription Lists
3. Create a new reader account using an email address that is NOT subscribed to your ESP.
4. Go through the checkout flow via the checkout block (either making a donation or purchasing a product via the checkout button block)
5. In the thank you modal, post-checkout, confirm there is no newsletter signup form
6. Go back to the engagement reader activation wizard and enable Newsletter Subscription Lists, and enable the lists set up in step 1.
7. Repeat checkout, but this time confirm the newsletter signup form is present in the thank you modal. Also confirm the newsletter list associated with the membership plan is not present
8. Sign up for the non-membership newsletter list via the signup form
9. Repeat checkout one more time, this time verifying neither the list associated with a membership plan as well as the list not associated with a membership plan are present in the form. (If you only had these two lists set up, then there should be no form)

### Other information:

* [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully ran tests with your changes locally?

<!-- Mark completed items with an [x] -->

chickenn00dle committed:
feat(ras-acc): update newsletter subscription lists setting copy

chickenn00dle committed:
fix: filter out already subscribed lists

chickenn00dle committed:
feat(ras-acc): filter out premium lists

adekbadek commented on includes/reader-activation/class-reader-activation.php:
```diff
@@ -405,6 +405,68 @@ public static function is_esp_configured() {
+		}
+
+		foreach ( $available_lists as $list_id => $list ) {
+			// Skip any lists the reader is already signed up for.
```
At the last testing step, I _do_ see the list I've signed up for. But I assume this is because I'm using MC with double opt-in.

chickenn00dle replied:
Hmm. To clarify, did the account you use for testing successfully complete the double opt-in process and get added to the audience in MC? I believe the `Newspack_Newsletters_Subscription::get_contact_lists` should return an array of list IDs the provided email is added to regardless of double or single opt in.

adekbadek replied:
> did the account you use for testing successfully complete the double opt-in process

No. 

> should return an array of list IDs the provided email is added to regardless of double or single opt in.

In my testing at least, the contact is not added in MC until they confirm the opt-in via email. That's the purpose of the double opt-in: allow only the email owner to sign up. 

Anyway, disregard my comment, it only muddles things up. This PR is not about the `Newspack_Newsletters_Subscription::get_contact_lists` method.

adekbadek reviewed (approved):
Maybe unrelated to these changes – let me know if I should add an issue on Asana – when initiating the donation flow as an anonymous reader, registration is the first step. After registering, and clicking "Continue", the modal is closed – and the donation flow is not continued. 

Other than that, works as expected.

chickenn00dle committed:
fix: update wc memberships subscription list helper method name

chickenn00dle commented:
Just noting a small change I've added after the approval. I've updated the method name for `is_membership_list` since this has been changed in one of the partner PRs to this one: https://github.com/Automattic/newspack-plugin/pull/3105/commits/400dd4135c1d77c80f5f8413f521cd3b9bd1b531
